content/blog/2019-12-13-futil-fsm/index.md

```diff
@@ -0,0 +1,315 @@
+     (new-std y0 (std_reg 32 0))
+     (-> (@ const3 out) (@ y0 in))
+     (new-std const3 (std_const 32 2))
+     (new-std z0 (std_reg 32 0))
+     (-> (@ const4 out) (@ z0 in))
+     (new-std const4 (std_const 32 4)))
+==> (seq
+     (par
+      (enable a0 const0)
+      (enable b0 const1))
+     (if (@ gt0 out) (enable gt0 a0 const2)
+         (enable y0 const3)
+         (enable z0 const4)))))
+```
+
+The first arrow is pointing to the structure and the second arrow is pointing to the control.
+The structure is an unordered list of two kinds of simple statements: `(new-std b0 (std_reg 32 0))` stands for instantiation of the library component `b0` with bitwidth parameter `32` and value parameter of `0`, and `(-> (@ a0 out) (@ gt0 left))` represents wiring the `out` port of component `a0` with the `left` port of component `gt0`. The control part specifies which components are *active* with `enable` keyword, and the execution logic with `par`, `seq`,`if` and`while` keywords. Think of activating a component like a function call. When a component is *active* it is allowed to *run* and produce valid outputs.
+
+In this project, we are interested in changing all the control logic to finite state machines (FSMs) and then generate simulatable Verilog program based on both FSMs and the structures.
+
+
+
+## Design Overview
+
+Futil is the backend for Dahlia. The Futil semantics are designed to allow for easy translation from higher level language, like Dahlia, but this creates a gap between the Futil semantics and Verilog implementations. The table below shows the efforts required to translate the Futil semantics to synthesizable Verilog implementations.
+
+| FuTIL Semantics   | Verilog         |
+| ----------------- | --------------- |
+| Invalid wire      | Read wires      |
+| Component Reusing | MUX, Read wires |
+| Control           | FSM             |
+
+### *Read* Signals
+
+In FuTIL semantics, `enable` keyword is used to determine whether a component is active. It is the easiest way of translating a program into hardware. However, this implicitly assumes that the signal on a wire is not valid or readable until we `enable` a component. We therefore require any *data* wire to have one extra bit to specify whether the signal is readable.
+
+### *MUX*
+
+A component can be used more than once in FuTIL. For instance, if we reuse variable `a0`, then the register need to choose the input from const0 and const2. This introduces a multiplexer (MUX).
+
+```futil
+(enable a0 const0)
+(enable a0 const2)
+```
+
+At different time step, read signals tells which wire to the MUX is readable. Therefore, read wires serve as **sel** signals for MUX.
+
+### *FSM*
+
+In FuTIL, there are control logics like `if`, `while` and etc. This can be translated into FSM in Verilog implementation, which is the main goal of this project. However, before getting to that, we created intermediate FSM expressions in FuTIL. An FSM component has:
```

Comment:
- "control logics" -> "control constructs"
- "and etc." -> just "etc."
- "This" -> "These" (plural)